### ISSUE TITLE
Unable to mount VHD images on dosbox-x

### ISSUE THREAD

**Issue by marcelofrau** [joncampbell123/dosbox-x#3987]
### Question

I've tried to create an VHD image with different tools (VHDManager and VirtualBox), in both cases I can attach/mount the vhd to my host normally, but when I try to use with dosbox-x, it doesn't work. (In all my attempts I unmount/detach the VHD first so the file is not being in use/locked by the OS)

I am following the documentation: https://dosbox-x.com/wiki/Guide%3AManaging-image-files-in-DOSBox%E2%80%90X

But whenever I try to IMGMOUNT the vhd image, it gives me the error: 
![image](https://user-images.githubusercontent.com/1341820/216816762-dc86a54c-8359-4d3c-a48b-c2136c817e22.png)

I have made a recording of my screen attempting to mount and attached the console output below:
![Code_DR2I52pA5I](https://user-images.githubusercontent.com/1341820/216816707-a1a523bc-4fc0-4823-94ec-e2524783722b.gif)

And even with the console, I cannot find what can be the problem. 
Below you can find the console from dosbox-x when I try to imgmount the vhd file.

```
LOG: Early LOG Init complete
LOG: Win32 EnumDisplayDevices #0: name=\\.\DISPLAY1 string=NVIDIA GeForce GTX 1060 3GB
LOG: Win32 EnumDisplayDevices #1: name=\\.\DISPLAY2 string=NVIDIA GeForce GTX 1060 3GB
LOG: Win32 EnumDisplayDevices #2: name=\\.\DISPLAY3 string=NVIDIA GeForce GTX 1060 3GB
LOG: Win32 EnumDisplayDevices #3: name=\\.\DISPLAY4 string=NVIDIA GeForce GTX 1060 3GB
LOG: DOSBox-X's working directory: E:\PC\_emulation\win98
LOG: Logging init: beginning logging proper. This is the end of the early init logging
LOG: Logging: No logfile was given. All further logging will be discarded.
LOG: Win32 EnumDisplayDevices #0: name=\\.\DISPLAY1 string=NVIDIA GeForce GTX 1060 3GB
LOG: Win32 EnumDisplayDevices #1: name=\\.\DISPLAY2 string=NVIDIA GeForce GTX 1060 3GB
LOG: Win32 EnumDisplayDevices #2: name=\\.\DISPLAY3 string=NVIDIA GeForce GTX 1060 3GB
LOG: Win32 EnumDisplayDevices #3: name=\\.\DISPLAY4 string=NVIDIA GeForce GTX 1060 3GB
LOG: DOSBox-X version 2022.12.26 (Windows SDL1)
LOG: Windows keyboard layout ID is 0x0416
LOG: Host keyboard layout is now us (US English)
LOG: Mapper keyboard layout is now us (US English)
LOG: The default output for the video system: direct3d
LOG: Configured windowposition: -
LOG: SDL: Windows 11 round corners will be disabled.
LOG: Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (508.000 x 285.750 mm) (20.000 x 11.250 in) (96.000 x 96.000 DPI)
LOG: ISA BCLK: 8333333.333Hz (25000000/3)
LOG: monopal: green,
LOG: Active save slot: 1 [Empty]
LOG: USING AVI+ZMBV
LOG: Max 3670016 sz 524288
LOG: Final 524288
LOG: MIDI:Opened device:win32
LOG: Pentium CMPXCHG8B emulation is enabled
LOG: One joystick reported, initializing with 4axis
LOG: Using joystick XInput Controller 0 with 6 axes, 10 buttons and 1 hat(s)
LOG: VOODOO LFB now at d0000000
LOG: Serial1: BASE 3f8h
LOG: Serial2: BASE 2f8h
LOG: Parallel1: BASE 378h
LOG: MPU-401 Registering I/O ports as if IBM PC MPU-401 at base 330h
LOG: Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (508.000 x 285.750 mm) (20.000 x 11.250 in) (96.000 x 96.000 DPI)
LOG: Allocated APM BIOS pm entry point at f000:ce40
LOG: Writing code to fce40
LOG: Writing code to fce60
LOG: ETHERNET: Backend not supported in this build: slirp
LOG: NE2000: Failed to open Ethernet backend slirp
LOG: Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (508.000 x 285.750 mm) (20.000 x 11.250 in) (96.000 x 96.000 DPI)
LOG: Windows: IID_ITaskbarList3 is available
LOG: ISA Plug & Play BIOS enabled
LOG: VGA ROM BIOS init callback
LOG: Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (508.000 x 285.750 mm) (20.000 x 11.250 in) (96.000 x 96.000 DPI)
LOG: Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (508.000 x 285.750 mm) (20.000 x 11.250 in) (96.000 x 96.000 DPI)
LOG: pixratio 1.000, dw false, dh false
LOG: Aspect ratio: 640 x 480  xToY=1.333 yToX=0.750
LOG: Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (508.000 x 285.750 mm) (20.000 x 11.250 in) (96.000 x 96.000 DPI)
LOG: Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (508.000 x 285.750 mm) (20.000 x 11.250 in) (96.000 x 96.000 DPI)
LOG: Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (508.000 x 285.750 mm) (20.000 x 11.250 in) (96.000 x 96.000 DPI)
LOG: Aspect ratio: 640 x 480  xToY=1.333 yToX=0.750
LOG: Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (508.000 x 285.750 mm) (20.000 x 11.250 in) (96.000 x 96.000 DPI)
LOG: WARNING: No translation support (to host) for code page 0
LOG: XMS: 50 handles allocated for use by the DOS environment
LOG: EMS page frame at 0xe000-0xefff
LOG: COMMAND.COM env size:             720 bytes
LOG: COMMAND.COM environment block:    0xcc01 sz=0x002d
LOG: COMMAND.COM main body (PSP):      0xcc2f sz=0x009a
LOG: COMMAND.COM stack:                0xcc49
LOG: pixratio 1.350, dw false, dh false
LOG: Aspect ratio: 720 x 540  xToY=1.333 yToX=0.750
LOG: Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (508.000 x 285.750 mm) (20.000 x 11.250 in) (96.000 x 96.000 DPI)
LOG: Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (508.000 x 285.750 mm) (20.000 x 11.250 in) (96.000 x 96.000 DPI)
LOG: Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (508.000 x 285.750 mm) (20.000 x 11.250 in) (96.000 x 96.000 DPI)
LOG: IDE: index 0 slave=0
LOG: Aspect ratio: 720 x 540  xToY=1.333 yToX=0.750
LOG: Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (508.000 x 285.750 mm) (20.000 x 11.250 in) (96.000 x 96.000 DPI)
```

### Have you checked that no similar question(s) exist?

- [X] I have searched and didn't find any similar question.

### Code of Conduct & Contributing Guidelines

- [X] I agree to follow the code of conduct and the contributing guidelines.

**Comment by joncampbell123** [comment]
As far as I can tell, the most basic VHD format is a flat image, but there are other types as well: https://en.wikipedia.org/wiki/VHD_(file_format)

Are you using anything like a dynamic image or perhaps differencing images?

Is the image the traditional VHD format or the newer VHDX?

**Comment by marcelofrau** [comment]
It's fixed and traditional VHD. I tried to create the image with VirtualBox and with VHD Manager, both of them offered the new format VHDX or dynamic but I have chosen fixed and the old VHD format.

If it is a flat file, do you know if can I rename it back and fort to .IMG? then I can use this as a workaround for my usecase.

**Comment by maron2000** [comment]
You can mount your image by using device number, e.g. `imgmount 2 hdd.vhd`.
Also partioning the disk by `fdisk` works but `format` didn't work.
Maybe some problem in accessing the FAT?
I'm not so sure though.

**Comment by rderooy** [comment]
You can only format a volume, once it has been assigned a drive letter. Mounting by device-number does not assign a drive letter, hence format fails.

Once the volume is partitioned with `fdisk`, try if you can do a `imgmount c hdd.vhd`.

**Comment by marcelofrau** [comment]
Ok, let me try, I'll format and make sure the volume was not already mounted. This was the same steps I mentioned in the first message. The VHD is not mounted on other place and was dedicated to be mounted only on dosbox-x or exclusively mounted on only one place.

**Comment by marcelofrau** [comment]
![image](https://user-images.githubusercontent.com/1341820/217529330-ca9da8ef-cf01-4459-be90-b36df720c64f.png)

I am still facing the same issue. VHD is just formatted as FAT and not mounted anywhere else.
The command you mentioned @rderooy is not working :(

**Comment by maron2000** [comment]
@rderooy  Thanks for your information.
So after some trial, I found a way to mount vhd files.
First, I booted a Windows98 boot disk and ran `fdisk` and `format`.
![fdisk_win98](https://user-images.githubusercontent.com/68574602/217532933-e82aa2c9-12aa-482c-8abf-56f99ca7607e.png)

Then I mounted the formatted vhd file. It failed to detect the drive geometry so I designated it by manual.
You can find out the geometry by mounting the image by device number. 
(see the screenshot below)

![imgmount](https://user-images.githubusercontent.com/68574602/217533262-ec192064-3a30-4701-9baa-552c922a754f.png)

Hope it helps.

**Comment by rderooy** [comment]
Yeah, I somewhat remember playing with this 2 years ago or so, and had similar issues. VHD support is very limited.

Normally on real DOS you first need to have a HDD partitioned before DOS will assign a drive letter. You may recall that FDISK will force you to reboot after creating partitions because the drive letter assignment is done earlier. Once a drive letter is assigned, you can format it in regular DOS. This does NOT work with the integrated DOS that DOSBox-X provides.

So your only option if you really want to use VHD right now is to do:
```
imgmount 2 myhdd.vhd
boot 2
```
This assumes myhdd.vhd is partitioned, formatted and has been made bootable.

**Comment by rderooy** [comment]
@maron2000 I did not know you had to give the geometry. I figured that data would be in the VHD header.

**Comment by maron2000** [comment]
@rderooy I don't know why but I had to give the geometry in order to mount. So strange since you don't need it if you mount by device numbers.

![command_001](https://user-images.githubusercontent.com/68574602/217537387-cddf3af0-c2a5-40fe-b0ec-66402232c2a3.png)

**Comment by rderooy** [comment]
If you mount by device number, it does not try to parse the partition table and filesystem. Which it does do when you mount by drive letter.

**Comment by maron2000** [comment]
@rderooy Yes, you're right, but the correct geometry is recognized even when mounted by device numbers, which is why I felt it strange that it fails to detect it when mounting by drive letter.
![imgmount2](https://user-images.githubusercontent.com/68574602/217540287-a07d3706-ad06-4d22-8e88-e4d3080bc5b5.png)

**Comment by rderooy** [comment]
Looking around. The original VHD format (aka VPC or VirtualPC format) is apparently little different from a RAW format.

You can create one on Linux as such:
```
qemu-img create -f vpc test.vpc 40M -o subformat=fixed
```
After this, I can indeed mount it in DOSBox-X with:
```
imgmount 2 test.vpc
```
The DOSBox-X log will show:
```
LOG: IDE: index 0 slave=0
LOG: VHD image detected: 512,17,5,964
LOG: Mapping BIOS DISK C/H/S 964/5/17 as IDE 964/5/17
```
Obviously mounting it as C: at this point will fail due to a lack of partition and filesystem data.

I then mounted a DOS 6.22 disk on A: and booted from it.
```
imgmount a Disk1.img
boot a:
```
Once DOS 6.22 booted, I exited the installer, and ran `fdisk` to create a primary partition. Fdisk rebooted and I once again booted into DOS 6.22, again exited the installer, and ran `format c: /s`.

I now exited dosbox-x and started it as such:
```
dosbox-x -c 'imgmount c test.vpc' -c 'boot c:'
```
![guest os_000](https://user-images.githubusercontent.com/17701098/217649695-7125f3e7-2343-4167-832c-cb3548e6149e.png)

**Comment by maron2000** [comment]
I tried changing the file extension from `.vhd` to `.vpc` and it worked.
Too strange to understand LOL.

![command_000](https://user-images.githubusercontent.com/68574602/217680492-667f1e4b-e619-4adc-9642-b12e5b00ce54.png)

**Comment by rderooy** [comment]
Looking at the `imgmount` code in `src/dos/dos_programs.cpp`, thinks are a bit confusing.

It seems there are two cases, both of which will cause an identical "VHD image detected..." message to be printed to the log.

The first in the **MountFat()** function, https://github.com/joncampbell123/dosbox-x/blob/791b732344c616b5a9447f24fcb8a6744c88e0b3/src/dos/dos_programs.cpp#L5741 
Here it tries to use `src/ints/bios_vhd.cpp` to open the file, if the filename ends in VHD (upper or lower case).
```cpp
//load the file with imageDiskVHD, which supports fixed/dynamic/differential disks
imageDiskVHD::ErrorCodes ret = imageDiskVHD::Open(ro?paths[i].c_str()+1:paths[i].c_str(), ro||roflag, &vhdImage);
```

The second case is in the **DetectGeometry()** function a bit further down https://github.com/joncampbell123/dosbox-x/blob/791b732344c616b5a9447f24fcb8a6744c88e0b3/src/dos/dos_programs.cpp#L5986
Here it tries to handle it itself. Here it also looks for the "conectix" signature in the footer that `imgmake` generates. This does not seem to be checking for a specific filename extension. So this must be the case that gets used when using an extension such as `.vpc`.

**Comment by maron2000** [comment]
I was preparing a PR to cancel falling back to DetectGeometry() which seems to work even with `.vhd` extension.
However, @joncampbell123 commited some fix, so I'm pending to put  the PR.
You can try the latest nightly build.

**Comment by joncampbell123** [comment]
I want to point out at least on my Linux box that qemu-img convert doesn't write a valid fixed size VHD image. The *.vhd file it writes seems to be missing the "conectix" footer at the end of the image.

The only way I have to test the VHD code is to hop onto my Intel mac or Windows laptop and make up some images from VirtualBox at this time. The VHD images it made seem to work fine with DOSBox-X. The fixes last night help to identify that it saw a VHD image even if the image is dynamic.

**Comment by maron2000** [comment]
@joncampbell123 Thanks for your comment.
So I rearranged my PR that I mentioned above based on the latest commit, which should fix this issue (#3996).
You no longer have to rename your image or give parameters manually.

**Comment by grapeli** [comment]
I created vhd images with `qemu-img` (version 7.2.50) without any problems and they work fine in DOSBox-X.

Dynamic 8GB image with bootable FreeDOS.
[dynamic.vhd.zip](https://github.com/joncampbell123/dosbox-x/files/10709484/dynamic.vhd.zip)

`qemu-img create -f vpc dynamic.vhd 8G -o subformat=dynamic,force_size=off`

Then I partitioned it and formatted it. All this was done in DOSBox-X under FreeDOS.

`dosbox-x -c 'imgmount c /tmp/dynamic.vhd' -c 'boot -l c'`

Fixed size.
`qemu-img create -f vpc /tmp/test.vhd 40M -o subformat=fixed,force_size=off`

**Comment by rderooy** [comment]
FYI, there is also support for mounting qcow2 images, but for some reason you can only mount them on a drive number, not on a drive letter.

**Comment by rderooy** [comment]
And I think `imgmake` should be updated to allow mixed case extensions for creating VHD images. Right now the line in `dos_programs.cpp` reads:
```
 if((mediadesc == 0xF8) && (temp_line.find(".vhd")) != std::string::npos) {
```
Which only matches on lower case.

**Comment by joncampbell123** [comment]
> I created vhd images with `qemu-img` (version 7.2.50) without any problems and they work fine in DOSBox-X.
> 
> Dynamic 8GB image with bootable FreeDOS.
> [dynamic.vhd.zip](https://github.com/joncampbell123/dosbox-x/files/10709484/dynamic.vhd.zip)
> 
> `qemu-img create -f vpc dynamic.vhd 8G -o subformat=dynamic,force_size=off`
> 
> Then I partitioned it and formatted it. All this was done in DOSBox-X under FreeDOS.
> 
> `dosbox-x -c 'imgmount c /tmp/dynamic.vhd' -c 'boot -l c'`
> 
> Fixed size.
> `qemu-img create -f vpc /tmp/test.vhd 40M -o subformat=fixed,force_size=off`
> 
> 

Maybe one weekend programming task I could try is to update imgmake to support making VHD and QCOW images in addition to raw flat images.

**Comment by maron2000** [comment]
@rderooy I fixed the `imgmake` code to make file extension (`.vhd`) check case insensitive, and added to my PR.

**Comment by maron2000** [comment]
I added codes to mount qcow2 images by drive letters.
I just copied the code to mount it by device numbers, so it's experimental and there maybe some flaws.
I prepared the image by the following command, and formatted by booting Win98 boot disk.
`qemu-img create -f qcow2 -o compat=0.10 hdd.qcow2 40M`

![command_000](https://user-images.githubusercontent.com/68574602/218268898-3001ea61-a1c8-4daf-9882-bb725d7d52bb.png)
